File: backend/CattleTrace/api/v1/serializers/slaughter.py
"""Slaughter record serializers."""

from rest_framework import serializers

from CattleTrace.models import Abattoir, Animal, SlaughterRecord, User


class AbattoirSerializer(serializers.ModelSerializer):
    class Meta:
        model = Abattoir
        fields = (
            'id',
            'name',
            'license_no',
            'county',
            'address',
            'contact',
            'is_active',
        )


class SlaughterRecordSerializer(serializers.ModelSerializer):
    inspector_name = serializers.SerializerMethodField()
    dressing_percentage = serializers.DecimalField(
        max_digits=5,
        decimal_places=2,
        read_only=True,
    )
    animal_tag = serializers.CharField(source='animal.tag_number', read_only=True)
    abattoir_detail = AbattoirSerializer(source='abattoir', read_only=True)

    class Meta:
        model = SlaughterRecord
        fields = (
            'id',
            'animal',
            'animal_tag',
            'abattoir',
            'abattoir_detail',
            'slaughter_date',
            'slaughter_no',
            'batch_number',
            'live_weight_kg',
            'carcass_weight_kg',
            'dressing_percentage',
            'hide_weight_kg',
            'offal_weight_kg',
            'inspector',
            'inspector_name',
            'inspection_result',
            'condemnation_reason',
            'meat_grade',
            'notes',
            'created_at',
        )
        read_only_fields = ('id', 'inspector', 'dressing_percentage', 'created_at')

    def get_inspector_name(self, obj):
        if obj.inspector:
            return obj.inspector.get_full_name() or obj.inspector.username
        return None

    def validate_animal(self, animal):
        if animal.status != Animal.Status.ALIVE:
            raise serializers.ValidationError(
                f"Cannot create a slaughter record for an animal with status "
                f"'{animal.get_status_display()}'. Only Active animals may be slaughtered."
            )
        return animal

    def create(self, validated_data):
        user = self.context['request'].user
        if user.role in (User.Role.INSPECTOR, User.Role.ABATTOIR, User.Role.ADMIN):
            validated_data['inspector'] = user
        return super().create(validated_data)

File: frontend/src/app/lib/api/slaughter.ts
import type { SlaughterRecord } from "../types";
import type { PaginatedResponse } from "./animals";

export interface ApiAbattoir {
  id: number;
  name: string;
  license_no: string;
  county: string;
  address: string;
  contact: string;
  is_active: boolean;
}

export interface ApiAbattoirPayload {
  name: string;
  license_no: string;
  county: string;
  address?: string;
  contact?: string;
  is_active?: boolean;
}

export type ApiInspectionResult = "passed" | "passed_partial" | "condemned";

export interface ApiSlaughterRecord {
  id: number;
  animal: number;
  animal_tag: string;
  abattoir: number | null;
  abattoir_detail: ApiAbattoir | null;
  slaughter_date: string;
  slaughter_no: string;
  batch_number: string;
  live_weight_kg: string;
  carcass_weight_kg: string;
  dressing_percentage: string | null;
  hide_weight_kg: string | null;
  offal_weight_kg: string | null;
  inspector: number | null;
  inspector_name: string | null;
  inspection_result: ApiInspectionResult;
  condemnation_reason: string;
  meat_grade: string;
  notes: string;
  created_at: string;
}

export interface SlaughterRecordListParams {
  page?: number;
  pageSize?: number;
  ordering?: "slaughter_date" | "-slaughter_date" | "created_at" | "-created_at";
}

export interface SlaughterStats {
  total: number;
  this_week: number;
  today: number;
  this_month: number;
  verified: number;
  compliance_rate: number;
}

export type SlaughterRecordListResponse = PaginatedResponse<ApiSlaughterRecord> | ApiSlaughterRecord[];

const inspectionVerified: Record<ApiInspectionResult, boolean> = {
  passed: true,
  passed_partial: true,
  condemned: false,
};

export function mapApiSlaughterRecordToSlaughterRecord(record: ApiSlaughterRecord): SlaughterRecord {
  return {
    id: String(record.id),
    animalId: String(record.animal),
    animalRfid: record.animal_tag,
    abattoirId: record.abattoir ? String(record.abattoir) : "",
    abattoirName: record.abattoir_detail?.name ?? "Unknown abattoir",
    chainNumber: record.slaughter_no,
    batchNumber: record.batch_number || undefined,
    carcassId: record.slaughter_no,
    slaughterDate: record.slaughter_date,
    liveWeightKg: record.live_weight_kg ? parseFloat(record.live_weight_kg) : undefined,
    carcassWeightKg: record.carcass_weight_kg ? parseFloat(record.carcass_weight_kg) : undefined,
    dressingPct: record.dressing_percentage ? parseFloat(record.dressing_percentage) : undefined,
    inspectionResult: record.inspection_result,
    meatGrade: record.meat_grade || undefined,
    inspectorName: record.inspector_name || undefined,
    feedback: record.condemnation_reason || record.notes || undefined,
    verified: inspectionVerified[record.inspection_result] ?? false,
  };
}
